transport/http/outbound.go

```diff
@@ -41,46 +45,50 @@ var (
 	errOutboundNotStarted     = errors.ErrOutboundNotStarted("http.Outbound")
 )
 
-type outboundConfig struct {
-	keepAlive time.Duration
-}
+// NewOutbound builds a new HTTP outbound that sends requests to the given
+// URL.
+//
+// Deprecated: create outbounds through NewPeerListOutbound instead
+func NewOutbound(urlStr string, opts ...AgentOption) transport.Outbound {
+	agent := NewAgent(opts...)
 
-var defaultConfig = outboundConfig{keepAlive: 30 * time.Second}
+	urlTemplate, hp := parseURL(urlStr)
 
-// OutboundOption customizes the behavior of an HTTP outbound.
-type OutboundOption func(*outboundConfig)
+	peerID := hostport.PeerIdentifier(hp)
+	peerList := peerlist.NewSingle(peerID, agent)
 
-// KeepAlive specifies the keep-alive period for the network connection. If
-// zero, keep-alives are disabled.
-//
-// Defaults to 30 seconds.
-func KeepAlive(t time.Duration) OutboundOption {
-	return func(c *outboundConfig) {
-		c.keepAlive = t
-	}
+	peerList.Start()
+
+	return NewPeerListOutbound(peerList, urlTemplate)
 }
 
-// NewOutbound builds a new HTTP outbound that sends requests to the given
-// URL.
-func NewOutbound(url string, opts ...OutboundOption) transport.Outbound {
-	cfg := defaultConfig
-	for _, o := range opts {
-		o(&cfg)
+func parseURL(urlStr string) (*url.URL, string) {
+	parsedURL, err := url.Parse(urlStr)
+
+	if err != nil {
+		return nil, urlStr
 	}
 
-	// Instead of using a global client for all outbounds, we use an HTTP
-	// client per outbound if unspecified.
-	client := buildClient(&cfg)
+	return parsedURL, parsedURL.Host
+}
 
-	// TODO: Use option pattern with varargs instead
-	return &outbound{Client: client, URL: url, started: atomic.NewBool(false)}
+// NewPeerListOutbound builds a new HTTP outbound built around a PeerList
+// for getting potential downstream hosts.  The PeerList.ChoosePeer function
+// needs to return a *hostport.Peer or it will break assumptions we make in the
+// Call func
```

Comment:
The documentation doesn't need to talk about assumptions. Just follow [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

> PeerList.ChoosePeer MUST return *hostport.Peer objects.

It's implicit that anything else is invalid usage and behavior is undefined.